lychee-bin/src/commands/check.rs

```diff
@@ -16,98 +19,126 @@ use crate::{
 };
 use lychee_lib::{Client, Request, Response};
 
-pub(crate) async fn check<S>(
+pub(crate) async fn check(
     client: Client,
     cache: Arc<Cache>,
-    requests: S,
-    cfg: &Config,
-) -> Result<(ResponseStats, Arc<Cache>, ExitCode)>
-where
-    S: futures::Stream<Item = Result<Request>>,
-{
-    let (send_req, recv_req) = mpsc::channel(cfg.max_concurrency);
-    let (send_resp, mut recv_resp) = mpsc::channel(cfg.max_concurrency);
+    inputs: Vec<Input>,
+    cfg: Config,
+) -> Result<(Arc<RwLock<ResponseStats>>, Arc<Cache>, ExitCode)> {
+    let base = cfg.base.clone();
+    let verbose = cfg.verbose;
     let max_concurrency = cfg.max_concurrency;
-    let mut stats = ResponseStats::new();
-    let cache_ref = cache.clone();
 
+    let input_sources: HashSet<InputSource> = inputs.iter().map(|i| i.source.clone()).collect();
+
+    // Get handles for values that will be moved into the async closure
+    let cache_handle = cache.clone();
+    let stats = Arc::new(RwLock::new(ResponseStats::new()));
+    let stats_handle = stats.clone();
+    let pb = init_progress(cfg.no_progress);
+    let pb_handle = pb.clone();
+
+    let (sender, requests) =
+        Collector::from_chan(base, cfg.skip_missing, cfg.max_concurrency).await;
+
+    for input in inputs {
+        sender
+            .send(input)
+            .await
+            .expect("Cannot send input to channel");
+    }
+
+    // Explicitly drop the channel to stop the stream inside the collector
```

Comment:
You want to drop the `sender` once there's nothing to process anymore right? But won't `for_each_concurrent` hang forever because you didn't do that yet? (and can't, because you need the sender for more cloning)

I can repro a deadlock even with `time lychee --offline -b . '**/*.htm*' -T1` on an empty directory.

Reply:
Yeah, I'm beginning to think that there is no way to avoid a deadlock if I both the sender and the receiver have to be empty for the program to terminate. I had a longer discussion on the Tokio Discord about it and the conclusion was that I should use a `Semaphore` lock and "normal" recursion + `tokio_spawn` instead.

Reply:
ah im on their discord but didn't see you already were asking this a while ago

Reply:
Basically this was what we ended up with. Thoughts?

```rust
use std::sync::Arc;
use tokio::{
    sync::{OwnedSemaphorePermit, Semaphore},
    task::JoinHandle,
};

const MAX_CONCURRENCY: usize = 10;

fn recurse(permit: OwnedSemaphorePermit, i: usize) -> JoinHandle<()> {
    tokio::spawn(async move {
        handle_input(permit, i).await;
    })
}

async fn handle_input(permit: OwnedSemaphorePermit, i: usize) {
    println!("got = {i}");
    if i % 9 == 0 {
        recurse(permit, 10).await.unwrap();
    }
}

#[tokio::main]
async fn main() {
    let semaphore = Arc::new(Semaphore::new(MAX_CONCURRENCY));

    tokio::spawn(async move {
        for i in 0..10 {
            let permit = semaphore.clone().acquire_owned().await.unwrap();
            handle_input(permit, i).await;
        }
    })
    .await
    .unwrap();
}
```

Reply:
I tried this model on the branch, but it gets quite ugly because of ownership and `Send`. We'll probably need an `Arc` to pass some recursion state around (which encapsulates all the moving parts like the config, the client, and maybe even the progress bar).

Reply:
I think this is the way to go. I wish there was a higher-level primitive for this, `for_each_concurrent` seemed perfect.

What ownership problems do you have?

Reply:
Had to lock the state with an `Arc<RwLock<State>>`, but I guess I locked once too often. 😆 Let me push the branch and create a new PR so you can have a look if you like. 😉

Reply:
Here you go: https://github.com/lycheeverse/lychee/pull/489